### PR TITLE
fix(geo2ddatat): crop

### PR DESCRIPTION
**Pull request by likewei92**
Adding a function to crop 2d geodata

### CODE REVIEW

File: qmt/geometry/geo_2d_data.py [hunk 1]
```diff
@@ -132,10 +134,63 @@ def coord_list(self, part_name: str) -> List:
...
+            ind_name = name.split(":")[0] if ":" in name else name
+            cropped_part = crop_poly.intersection(part)
+            if cropped_part.is_empty:
+                continue
+            if isinstance(cropped_part, Polygon) or isinstance(
+                cropped_part, LineString
+            ):
+                cropped_geo.add_part(name, cropped_part)
+                ind_count[ind_name] += 1
+            elif isinstance(cropped_part, GeometryCollection):
```
Comment (CameronKing) [inline qmt/geometry/geo_2d_data.py]: At initialization, Geo2DData is intended to made from a Polygon or LineString, do we want to introduce other geometry types?

File: qmt/geometry/geo_2d_data.py [hunk 2]
```diff
@@ -132,10 +134,63 @@ def coord_list(self, part_name: str) -> List:
...
+        y_max = y_max_old if y_max is None else y_max
+        crop_poly = Polygon(
+            [[x_min, y_min], [x_min, y_max], [x_max, y_max], [x_max, y_min]]
+        )
+
+        ind_count = defaultdict(int)
+
+        for name, part in self.parts.items():
+            ind_name = name.split(":")[0] if ":" in name else name
+            cropped_part = crop_poly.intersection(part)
```
Comment (CameronKing) [inline qmt/geometry/geo_2d_data.py]: We should probably have a test for the case where a part edge is co-linear with the bounding box edge.

File: qmt/geometry/geo_2d_data.py [hunk 3]
```diff
@@ -132,10 +134,63 @@ def coord_list(self, part_name: str) -> List:
...
+            if cropped_part.is_empty:
+                continue
+            if isinstance(cropped_part, Polygon) or isinstance(
```
Comment (basnijholt) [inline qmt/geometry/geo_2d_data.py]: `isinstance(cropped_part, Polygon, LineString)` is the same

File: qmt/geometry/geo_3d_data.py [hunk 4]
```diff
@@ -20,8 +20,8 @@ class Geo3DData(GeoData):
         - build_order is a list of strings indicating the construction order
     """
 
-    def __init__(self, lunit: str = "nm"):
-        super().__init__(lunit)
+    def __init__(self, lunit: Optional[str] = None):
+        super().__init__("nm" if lunit is None else lunit)
```
Comment (basnijholt) [inline qmt/geometry/geo_3d_data.py]: `lunit or "nm"` is the same and better

File: qmt/geometry/freecad/objectConstruction.py [hunk 5]
```diff
@@ -97,7 +97,7 @@ def build(opts):
 
     """
     doc = FreeCAD.ActiveDocument
-    geo = Geo3DData()
+    geo = Geo3DData(opts["lunit"] if "lunit" in opts else None)
```
Comment (basnijholt) [inline qmt/geometry/freecad/objectConstruction.py]: this is `opts.get("lunit", None)`

File: qmt/geometry/geo_2d_data.py (hunk 1 above)
Comment (likewei92) [reply qmt/geometry/geo_2d_data.py]: This is necessary because the intersection of 2 Polygon/LineString objects may not return a Polygon/LineString . Think about this:
untitled
the intersection of crop and area1 will be two polygons, and it's returned as a GeometryCollection. In this case we iterate through it and add each polygon (as area1:0 and area1:1)

File: qmt/geometry/geo_2d_data.py (hunk 2 above)
Comment (likewei92) [reply qmt/geometry/geo_2d_data.py]: I've tested this and it works. The edge is included in the crop as a LineString. I think if you're using it for bdg all LineStrings are ignored, but from a geometry point of view it might be good to keep it there (for e.g. if you're setting boundary conditions for TF then it matters)

Comment (likewei92) [reply qmt/geometry/geo_2d_data.py]: I'll write a proper test for it as Andrey suggests though

File: qmt/geometry/geo_3d_data.py (hunk 4 above)
Comment (likewei92) [reply qmt/geometry/geo_3d_data.py]: That's a good trick!

**Comment by likewei92** [comment]
@aeantipov I've added a test
@basnijholt I've made the changes you requested

**Comment by likewei92** [comment]
I added some asserts to make sure that the crop is within bounds

**Comment by likewei92** [comment]
@basnijholt do you mind taking a look again so I can merge? :)

File: qmt/geometry/geo_2d_data.py [hunk 6]
```diff
@@ -132,10 +134,69 @@ def coord_list(self, part_name: str) -> List:
...
+            cropped_part = crop_poly.intersection(part)
+            if cropped_part.is_empty:
+                continue
+            if isinstance(cropped_part, (Polygon, LineString)):
+                cropped_geo.add_part(name, cropped_part)
+                ind_count[ind_name] += 1
+            elif isinstance(cropped_part, GeometryCollection):
+                for part in cropped_part:
+                    if not (isinstance(part, Polygon) or isinstance(part, LineString)):
+                        print(type(part))
```
Comment (basnijholt) [inline qmt/geometry/geo_2d_data.py]: Is this print statement supposed to be here?

File: qmt/geometry/geo_2d_data.py [hunk 7]
```diff
@@ -132,10 +134,69 @@ def coord_list(self, part_name: str) -> List:
...
+            y_max,
+        ] == cropped_geo.compute_bb(), "Cropped geometry has wrong bounds"
+
+        return cropped_geo
+
     def plot(
         self,
         parts_to_exclude: Optional[Sequence[str]] = None,
-        line_width: float = 20.0,
+        line_width: float = None,
```
Comment (basnijholt) [inline qmt/geometry/geo_2d_data.py]: Looks like this should be `Optional[float]`.

File: qmt/geometry/geo_2d_data.py [hunk 8]
```diff
@@ -146,7 +207,7 @@ def plot(
         parts_to_exclude : Sequence[str]
             Part/edge names that won't be plotted (Default value = None)
         line_width : float
-            Thickness of lines (only for edge lines). (Default value = 20.0)
+            Thickness of lines (only for edge lines). (Default value = None)
```
Comment (basnijholt) [inline qmt/geometry/geo_2d_data.py]: Maybe write what `None` will do.

File: qmt/geometry/geo_2d_data.py [hunk 9]
```diff
@@ -132,10 +134,69 @@ def coord_list(self, part_name: str) -> List:
...
+            ind_name = name.split(":")[0] if ":" in name else name
+            cropped_part = crop_poly.intersection(part)
+            if cropped_part.is_empty:
+                continue
+            if isinstance(cropped_part, (Polygon, LineString)):
+                cropped_geo.add_part(name, cropped_part)
+                ind_count[ind_name] += 1
+            elif isinstance(cropped_part, GeometryCollection):
+                for part in cropped_part:
+                    if not (isinstance(part, Polygon) or isinstance(part, LineString)):
```
Comment (basnijholt) [inline qmt/geometry/geo_2d_data.py]: `isinstance(part, (Polygon, LineString))`

File: qmt/geometry/geo_2d_data.py [hunk 10]
```diff
@@ -132,10 +134,69 @@ def coord_list(self, part_name: str) -> List:
...
+                        print(type(part))
+                    cropped_geo.add_part(f"{ind_name}:{ind_count[ind_name]}", part)
+                    ind_count[ind_name] += 1
+            else:
+                raise RuntimeError(
+                    "Unknown intersection of type ({type(cropped_part)}) encountered "
```
Comment (basnijholt) [inline qmt/geometry/geo_2d_data.py]: I think this should be an f-string.

File: qmt/geometry/geo_2d_data.py (hunk 6 above)
Comment (likewei92) [reply qmt/geometry/geo_2d_data.py]: Oops :P
